src/spec/hdmf_common.hpp

```diff
@@ -0,0 +1,29 @@
+
+const std::string sparse = R"delimiter(
+{"groups":[{"data_type_def":"CSRMatrix","data_type_inc":"Container","doc":"A compressed sparse row matrix. Data are stored in the standard CSR format, where column indices for row i are stored in indices[indptr[i]:indptr[i+1]] and their corresponding values are stored in data[indptr[i]:indptr[i+1]].","attributes":[{"name":"shape","dtype":"uint","dims":["number of rows, number of columns"],"shape":[2],"doc":"The shape (number of rows, number of columns) of this sparse matrix."}],"datasets":[{"name":"indices","dtype":"uint","dims":["number of non-zero values"],"shape":[null],"doc":"The column indices."},{"name":"indptr","dtype":"uint","dims":["number of rows in the matrix + 1"],"shape":[null],"doc":"The row index pointer."},{"name":"data","dims":["number of non-zero values"],"shape":[null],"doc":"The non-zero values in the matrix."}]}]})delimiter";
+
+const std::string namespaces = R"delimiter(
+{"namespaces":[{"name":"hdmf-common","doc":"Common data structures provided by HDMF","author":["Andrew Tritt","Oliver Ruebel","Ryan Ly","Ben Dichter"],"contact":["ajtritt@lbl.gov","oruebel@lbl.gov","rly@lbl.gov","bdichter@lbl.gov"],"full_name":"HDMF Common","schema":[{"source":"base"},{"source":"table"},{"source":"sparse"}],"version":"1.8.0"}]})delimiter";
+
+void registerVariables(std::map<std::string, const std::string*>& registry)
+{
+  registry["base"] = &base;
+  registry["table"] = &table;
+  registry["sparse"] = &sparse;
+  registry["namespace"] = &namespaces;
+};
```

Comment:
If I understand this correctly, the main purpose of `registerVariables` function is to initialize the `std::map` at runtime because to work around the issue that we cannot initialize the ``std::map` at compile time. An alternative approach would be to use a combination of ``std::array`` ,``std::pair``, and `std::string_view`` instead here, which I believe would look like this:

```
constexpr std::string_view base = R"delimiter()delimiter";

constexpr std::array<std::pair<std::string_view, std::string_view>, 2> myVariables = {{
    {"greeting", base},
}};
```

This would not give the key-based access like  the ``std::map``, but would require iterating through the array, where each entry is then a key-value pair. However, looking at ``NWBFile::cacheSpecifications``, it seems that is all we need to do anyways., which would then look like this:

```
for (const auto& entry : myVariables) {
        std::cout << entry.first << ": " << entry.second << std::endl;
    }
```

The main reason I am proposing this as a possible alternative is becuase it would avoid the need for passing around function pointers (which may not as intuitive as passing a variable) and everything would be initialized at compile time. What do you think?

Reply:
I think that makes sense! I wasn't aware of `std::string_view` and that seems like a better option for this purpose.

The only issue with having `myVariables` as an array initialized at compile time is that we probably want `cacheSpecifications` to accept variable length arrays? I templated the function for size, that seemed the most straightforward to me. But some other options include converting it to a `std::vector` or passing in the array iterators

Reply:
> I templated the function for size, that seemed the most straightforward to me. But some other options include converting it to a `std::vector` or passing in the array iterators

I agree, using a template in this case seems like a good option; it still keeps everything at compile time and calling the function still works the same. 👍 

Copying to std::vector just adds an extra step and more work because I don't think there is a zero-copy option to turn an std::array into an std::vector. Using iterators would work, but would require handing in 2 variables (i.e., the start and the end iterator) and we'd lose the ability to enforce the data type (i.e., the iterator could point to a std::vector<int> for all we know) so there is a chance of misuse and runtime errors.